orttraining/orttraining/training_api/checkpoint.cc

```diff
@@ -428,9 +538,54 @@ Status FromCheckpointState(
     const CheckpointState& state, const PathString& checkpoint_path, const bool include_optimizer_state) {
   flatbuffers::FlatBufferBuilder builder(1024);
 
+  fbs::utils::ExternalDataWriter external_data_writer = nullptr;
+  std::optional<std::ofstream> external_data_stream;
+  if (state.has_external_data) {
+    auto data_path = ExternalCheckpointDataPath(checkpoint_path);
+    external_data_stream = std::ofstream(data_path, std::ios::binary);
+
+    ORT_RETURN_IF(external_data_stream->fail(), "Failed to create checkpoint's external data file: ",
+                  ToUTF8String(data_path));
+
+    // setup the data writer to write aligned data to external_data_stream
+    external_data_writer = [&external_data_stream](int32_t data_type, gsl::span<const uint8_t> bytes,
+                                                   uint64_t& offset) {
+      // for now align everything to 4 or 8 bytes. we can optimize this later if needed.
+      int32_t alignment = 4;
+
+      // TODO: Add more special-cased 8 byte types if needed.
+      if (data_type == ONNX_NAMESPACE::TensorProto_DataType_INT64 ||
+          data_type == ONNX_NAMESPACE::TensorProto_DataType_DOUBLE) {
+        alignment = 8;
+      }
+
+      int64_t pos = external_data_stream->tellp();
+
+      if (pos % alignment != 0) {
+        // 8 bytes of 0's so we can pad to alignment 8 in a single `write`
+        constexpr static const uint64_t zeros = 0;
+        int64_t padding = alignment - (pos % alignment);
+        // skipping validation of this write. doesn't matter if this or the 'real' write below fails. if this does the
+        // other will as well as nothing will clear the failure bit in the ofstream in between the calls.
+        external_data_stream->write(reinterpret_cast<const char*>(&zeros), padding);
+        pos += padding;
+      }
+
+      external_data_stream->write(reinterpret_cast<const char*>(bytes.data()), bytes.size());
+      // TODO: include error code/message for why it failed. requires platform specific code
+      ORT_RETURN_IF(external_data_stream->fail(), "Failed writing external checkpoint data.");
+
+      // temporary sanity check. remove when code is unit tested
+      assert(pos + int64_t(bytes.size()) == external_data_stream->tellp());
+
+      offset = pos;
+      return Status::OK();
+    };
```

Comment:
Split this out into a helper function that creates the external data writer so it can be used in FromCheckpointState as well